SourceryFramework/Sources/Parsing/FileParser.swift

```diff
@@ -318,6 +336,144 @@ extension FileParser {
+                if inheritance < start {
+                    return ""
+                }
+            }
+            var end = start
+            var openingAngles = 1
+            while openingAngles > 0 {
+                end = source.index(after: end)
+                guard end != source.endIndex else { return "" }
+                // Count closing bracket, but exclude closures, e.g. ->
+                if source[end] == ">", source[source.index(before: end)] != "-" {
+                    openingAngles -= 1
+                } else if source[end] == "<" {
+                    openingAngles += 1
+                }
+            }
+            
+            return String(source[source.index(after: start) ..< end])
+        }
+        return ""
+    }
+    
+    fileprivate func extractGenericsWhereClause(source: String) -> String {
+        if let start = source.range(of: "where")?.lowerBound {
+            return String(source[source.index(start, offsetBy: "where".count) ..< source.endIndex])
+        }
+        return ""
+    }
+    
+    fileprivate func extractGenericTypeName(source: String) -> String {
+        return String(source[..<(source.index(of: "<") ?? source.endIndex)])
+    }
+    
+    fileprivate func extractGenericTypeParameters(source: String) -> [GenericTypeParameter] {
+        return source.commaSeparated()
+            .filter { component in !component.isEmpty }
+            .map { parseGenericTypeParameter(source: $0) }
+    }
+    
+    fileprivate func parseGenericTypeParameter(source: String) -> GenericTypeParameter {
+        let source = source.trimmingCharacters(in: .whitespacesAndNewlines)
+        if source.contains("<") {
+            let declaration = extractGenericsDeclaration(source: source)
+            let typeName = extractGenericTypeName(source: source)
+            return GenericTypeParameter(typeName: TypeName(typeName), type: Type(name: typeName, genericTypeParameters: extractGenericTypeParameters(source: declaration)))
+        } else {
+            return GenericTypeParameter(typeName: TypeName(source), type: Type(name: source))
+        }
+    }
+    
+    fileprivate func extractGenericTypePlaceholders(source: String) -> [GenericTypePlaceholder] {
+        return source.commaSeparated()
+            .filter { component in !component.isEmpty }
+            .map { parseGenericTypePlaceholder(source: $0) }
+    }
+    
+    fileprivate func parseGenericTypePlaceholder(source: String) -> GenericTypePlaceholder {
+        if let constraintStart = source.index(of: ":") {
+            let name = String(source[source.startIndex ..< constraintStart]).trimmingCharacters(in: .whitespacesAndNewlines)
+            let constraints = String(source[source.index(after: constraintStart) ..< source.endIndex])
+                .components(separatedBy: "&")
+                .map { constraint in
+                    recursivelyParseGenericPlaceholderDeclaration(source: constraint)
```

Comment:
do we need to do that for each type? I think it's only allowed to use concrete type as a first type of composition and everything else would be protocols, i.e. `ViewController<A> & Presentable & Dismissible`. Probably would be a bit faster if we only ever do it on first element.

Reply:
Unfortunately, we do. Unit test in FileParserSpec "extracts generics correctly" on line 270 shows specific example where this is needed.